### PR TITLE
RSDEV-1153: Fix PyRAT pagination lag

### PR DESCRIPTION
**Pull request by rs-fraser**
> [!NOTE]
> Stacked on #984 (the dialog extraction). Until that merges, this diff includes its changes too; review only the last three commits here (the fix, the regression tests, and the edge cases).

### Problem

In the PyRAT dialog, changing the page or the page size showed results one step behind the controls. The handlers fetched imperatively right after a `setState`, so each request was built from the previous render's `page` and `rowsPerPage` values. Changing the page size or sort while on a later page also fired a request for a page that no longer existed.

### Fix

- Drive the fetch off a memoised query string instead of fetching imperatively from the event handlers, so every pagination change produces exactly one request carrying the new values.
- Reset to the first page during render when the filters or sort change, so the reset and the new query land in a single request.

### Testing
AWS build has `user1a` configured with Pyrat to view the fix: https://rsdev-1153-pyrat-pagination-0b7cb50a-2.researchspace.com
(pangolin85 as `sysadmin1` is configured showing the previous bug behaviour): https://pangolin8085.researchspace.com

New browser-mode regression suite following the PubChem four-artifact pattern (`PyratDialog.spec.tsx`, `PyratDialog.story.tsx`, MSW mocks, page object):

- Changing the page size refetches with the new size and renders that many rows.
- Changing the page requests the matching offset.
- Page-size and sort changes made on a later page reset to the first page in a single request.
- Additional edge cases around the above.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by rs-fraser** [comment]
@greptileai

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,661 @@
...
+      // biome-ignore lint/suspicious/noExplicitAny: initial biome migration
+      .catch((error: any) => {
+        handlePyratError(error);
+      });
+  }, []);
+
+  useEffect(() => {
+    pyrat
+      .get(
+        `locations?serverAlias=${serverAlias}&s=full_name:asc&k=building_id&k=full_name&type=building&=status=available`,
```
[the remark this answers is not in the document]
Comment (rs-fraser) [reply src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx:288]: Good catch, but this is pre-existing code: this PR's refactor moved it verbatim from `Pyrat.tsx` into `PyratDialog.tsx`, and the behaviour is unchanged from main. To keep this PR focused on the pagination bug, it will be tidied separately under RSDEV-1256, which documents this finding along with the others from this review.

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx [hunk 2]
```diff
@@ -0,0 +1,661 @@
...
+            { None: "" },
+          );
+
+          setFilterSpecial({
+            ...filterSpecial,
+            building_id: {
+              ...filterSpecial.building_id,
+              enumObj,
+            },
+          });
```
[the remark this answers is not in the document]
Comment (rs-fraser) [reply src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx:307]: Good catch, but this is pre-existing code: this PR's refactor moved it verbatim from `Pyrat.tsx` into `PyratDialog.tsx`, and the behaviour is unchanged from main. To keep this PR focused on the pagination bug, it will be tidied separately under RSDEV-1256, which documents this finding along with the others from this review.

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx [hunk 3]
```diff
@@ -0,0 +1,661 @@
...
+      </DialogActions>
+    </Dialog>
+  );
+}
+
+// biome-ignore lint/suspicious/noExplicitAny: initial biome migration
+function createTinyMceTable(selectedAnimals: any[]) {
+  const pyratTable = document.createElement("table");
+  pyratTable.setAttribute("data-tableSource", "pyrat");
+  pyratTable.style = "font-size: 0.7em";
```
[the remark this answers is not in the document]
Comment (rs-fraser) [reply src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx:611]: Good catch, but this is pre-existing code: this PR's refactor moved it verbatim from `Pyrat.tsx` into `PyratDialog.tsx`, and the behaviour is unchanged from main. To keep this PR focused on the pagination bug, it will be tidied separately under RSDEV-1256, which documents this finding along with the others from this review.

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx [hunk 4]
```diff
@@ -0,0 +1,661 @@
...
+                            PYRAT_URL = server.url;
+                            PYRAT_ALIAS = server.alias;
+                          }}
+                        >
+                          <ListItemText primary={server.alias} secondary={server.url} />
+                        </ListItemButton>
+                      </ListItem>
+                      <Divider />
+                    </>
+                  ))}
```
[the remark this answers is not in the document]
Comment (rs-fraser) [reply src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx:582]: Good catch, but this is pre-existing code: this PR's refactor moved it verbatim from `Pyrat.tsx` into `PyratDialog.tsx`, and the behaviour is unchanged from main. To keep this PR focused on the pagination bug, it will be tidied separately under RSDEV-1256, which documents this finding along with the others from this review.

File: src/main/webapp/ui/src/tinyMCE/pyrat/pageObjects/PyratDialogPage.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  dataRowCount(): number {
+    return this.dataRows().elements().length;
+  }
+
+  get rowsPerPageSelect(): Locator {
+    return page.getByRole("combobox", { name: /rows per page/i });
+  }
+
+  get nextPageButton(): Locator {
+    return page.getByRole("button", { name: /go to next page/i });
```
Comment (tchoi-rs) [inline src/main/webapp/ui/src/tinyMCE/pyrat/pageObjects/PyratDialogPage.ts]: Do these need to be regex?

File: src/main/webapp/ui/src/tinyMCE/pyrat/Filter.tsx [hunk 6]
```diff
@@ -180,10 +180,13 @@ export default function Filter({
     // biome-ignore lint/suspicious/noExplicitAny: initial biome migration
     [key: string]: any;
   }) {
+    const labelId = `${name}-label`;
```
Comment (tchoi-rs) [inline src/main/webapp/ui/src/tinyMCE/pyrat/Filter.tsx]: Use `useId()`

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.story.tsx [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+ * mounts the animal listing, which is where the pagination logic lives.
+ */
+export function PyratDialogStory(): React.ReactNode {
+  const [open, setOpen] = React.useState(true);
+  return (
+    <StyledEngineProvider injectFirst>
+      <ThemeProvider theme={createAccentedTheme(ACCENT_COLOR)}>
+        <ErrorBoundary>
+          <Alerts>
+            {/** biome-ignore lint/a11y/useButtonType: story-only trigger */}
```
Comment (tchoi-rs) [inline src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.story.tsx]: As the fix isn't that expensive, it should not be `biome-ignore`d

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx [hunk 8]
```diff
@@ -395,6 +400,13 @@ export function PyratListing({ serverAlias, setSelectedAnimals }: { serverAlias:
     }&${params.join("")}&s=${orderBy}:${order}`;
   }, [filterCounter, order, orderBy, page, rowsPerPage, tableHeaderCells]);
 
+  // makeQueryString encodes the query parameters, while filterCounter is the
+  // explicit apply trigger. Keep both dependencies because animal_type selects
+  // the animals/pups collection rather than appearing in the query string.
+  useEffect(() => {
+    fetchAnimals();
+  }, [filterCounter, makeQueryString]);
```
Comment (tchoi-rs) [inline src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx:408]: This breaks down in multiple-clicking, as each click would increment the `filterCounter` which would then trigger a re-request.

Reading the code I am not sure why `filterCounter` even exists so I think to fix the root cause, this component needs to be redesigned using TanStack Query to properly eliminate duplicate request issues.

**Review by tchoi-rs — CHANGES_REQUESTED** [review]

File: src/main/webapp/ui/src/tinyMCE/pyrat/pageObjects/PyratDialogPage.ts (hunk 5 above)
Comment (rs-fraser) [reply src/main/webapp/ui/src/tinyMCE/pyrat/pageObjects/PyratDialogPage.ts]: Nope. Matching on exact string now.

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.story.tsx (hunk 7 above)
Comment (rs-fraser) [reply src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.story.tsx]: Fixed properly

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx (hunk 8 above)
Comment (rs-fraser) [reply src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx:408]: Yeah although this is a pre-existing issue. I've created https://researchspace.atlassian.net/browse/RSDEV-1256 for those, or if you think the tanstack re-write deserves it's own ticket please create it.

File: src/main/webapp/ui/src/tinyMCE/pyrat/Filter.tsx (hunk 6 above)
Comment (rs-fraser) [reply src/main/webapp/ui/src/tinyMCE/pyrat/Filter.tsx]: done

File: src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx (hunk 8 above)
Comment (tchoi-rs) [reply src/main/webapp/ui/src/tinyMCE/pyrat/PyratDialog.tsx:408]: No problem with leaving it for another time.

**Review by tchoi-rs — APPROVED** [review]
